services/logger/zap.go

```diff
@@ -0,0 +1,215 @@
+	including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and / or sell copies of the Software, and to permit persons
+	to whom the Software is furnished to do so, subject to the following conditions:
+	The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+	WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+package logger
+
+// Merak's wrapper for the zap logger
+
+import (
+	"flag"
+	"fmt"
+	"log/syslog"
+	"os"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/tchap/zapext/zapsyslog"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var _ Logger = (*MerakLog)(nil)
+
+type MerakLog struct {
+	Zap   *zap.SugaredLogger
+	Level zap.AtomicLevel
+}
+
+type Level int8
+
+// The supported Log Levels
+const (
+	DEBUG Level = Level(zap.DebugLevel)
+	INFO  Level = Level(zap.InfoLevel)
+	WARN  Level = Level(zap.WarnLevel)
+	ERROR Level = Level(zap.ErrorLevel)
+	PANIC Level = Level(zap.PanicLevel)
+	FATAL Level = Level(zap.FatalLevel)
+)
+
+// Where to log
+type Location int
+
+const (
+	Syslog Location = iota + 1
+	File
+	Stdout
+)
+
+type options struct {
+	location    Location
+	level       zap.AtomicLevel
+	serviceName string
+	fileName    string
+}
+
+type merakLogError struct {
+	Err     error
+	Message string
+}
+
+func (e merakLogError) Error() string {
+	return fmt.Sprintf("%s: %v", e.Message, e.Err)
+}
+
+func newCore(opts *options) (any, error) {
+	config := zap.NewProductionEncoderConfig()
+	config.TimeKey = "time"
+	encoder := zapcore.NewJSONEncoder(config)
+	switch loc := opts.location; loc {
+	case Syslog:
+		flagTag := flag.String("app", opts.serviceName, "syslog tag")
+		writer, err := syslog.New(syslog.LOG_ERR|syslog.LOG_LOCAL0, *flagTag)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to set up syslog")
+		}
+		core := zapsyslog.NewCore(opts.level, encoder, writer)
+		return core, nil
+	case File:
+		f, err := os.Create(opts.fileName)
+		if err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("failed to write to file %s", opts.fileName))
+		}
+		core := zapcore.NewCore(encoder, zapcore.AddSync(f), opts.level)
+		return core, nil
+	case Stdout:
+		return zapcore.NewCore(encoder, os.Stdout, opts.level), nil
+	}
+	return nil, merakLogError{errors.New("invalid logger case"), strconv.Itoa(int(opts.location))}
+}
+
+// Creates a new logger that writes to stdout with the given log level
+func NewLogger(level Level) (*MerakLog, error) {
```

Comment:
@phudtran So, we need to decide which logger (stdout, syslog, or file) we want to use in each service? Let's make our logger saving the log data into /var/log/merak for each service. Can you write an example to show how to use this logger to write the log data into a file and still can retrieve the log info from `kubectl logs` command.

Reply:
Sure, I've updated the PR and changed the default logger to log both to /var/log/merak/service_name and stdout/console which will be visible from kubectl logs

Reply:
Also updated the PR with an example.